### PR TITLE
feat: add clippy

### PR DESCRIPTION
**Pull request by wispl**
Rust linting using cargo clippy, done to the best of my limited abilities. Since clippy lints the whole project, some filtering has to be done. I am not too keen on the recursion but sometimes there is useful information from the child messages. Thanks!

### CODE REVIEW

**Comment by wispl** [comment]
Fixes some issues when using with pedantic mode. There can be multiple spans and each one is related to its parent message. Apparently clippy also outputs a suggested replacement message, currently I have it displayed by surrounding it with parenthesis, which might not be the best of ideas, especially when an empty string means deletion. Suggestions?

File: lua/lint/linters/clippy.lua [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  help = vim.diagnostic.severity.HINT,
+}
+
+local function parse(diagnostics, file_name, item)
+  for _, span in ipairs(item.spans) do
+    if span.file_name == file_name then
+      local message = item.message
+      if span.suggested_replacement ~= vim.NIL then
+        message = message .. " (" .. tostring(span.suggested_replacement) .. ")"
+      end
```
Comment (mfussenegger) [inline lua/lint/linters/clippy.lua:13]:
```suggestion
      if span.suggested_replacement ~= vim.NIL then
        message = message .. "\nSuggested replacement:\n\n" .. tostring(span.suggested_replacement)
      end
```

Not sure if that's better. I can also merge it as is.

File: lua/lint/linters/clippy.lua [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+return {
+  cmd = "cargo",
+  args = { "clippy", "--message-format=json" },
+  stdin = false,
+  append_fname = false,
+  parser = function(output, bufnr)
+    local diagnostics = {}
+    local items = #output > 0 and vim.split(output, "\n") or {}
+    local file_name = vim.fn.expand("%")
```
Comment (mfussenegger) [inline lua/lint/linters/clippy.lua:40]:
```suggestion
    local file_name = vim.api.nvim_buf_get_name(bufnr)
```

Given that nvim-lint runs the linter async by the time the output gets parsed the active buffer could've changed, so this should use the `bufnr` to resolve the name.

**Comment by wispl** [comment]
Applied the suggestions, I think the newlines might be better. Thanks for reviewing!

**Pull request merged**